store/src/commonMain/kotlin/com/dropbox/external/store5/ConflictResolver.kt

```diff
@@ -0,0 +1,14 @@
+package com.dropbox.external.store5
+
+typealias GetLastFailedWriteTime<Key> = suspend (key: Key) -> Long?
+typealias SetLastFailedWriteTime<Key> = suspend (key: Key, datetime: Long?) -> Boolean
+typealias DeleteFailedWriteRecord<Key> = suspend (key: Key) -> Boolean
+
+/**
+ * Resolves conflicts among local and remote data sources.
+ */
+data class ConflictResolver<Key : Any, Input : Any, Output : Any>(
```

Comment:
How would i implement a different conflict resolution ?

Reply:
You wouldn't implement a different `ConflictResolver`. Currently the implementation of a conflict resolution system is deferred to the user and their remote service. Right now we try to prevent conflicts by eagerly pushing changes to server before fetching new data. But we only do this when there is a write request failure saved on disk or there is a write request in queue. This `ConflictResolver` class was originally named `ConflictResolution`. It doesn't actively resolve conflicts. It binds to a persistent data source and reads/writes latest request failure for a given `Key`.

Thoughts?

cc: @digitalbuddha

Reply:
See also: https://github.com/MobileNativeFoundation/Store/pull/450#discussion_r966444331

Reply:
Renamed to `Bookkeeper` in line with [offline-first guidance](https://developer.android.com/topic/architecture/data-layer/offline-first#conflict-resolution). See https://github.com/MobileNativeFoundation/Store/pull/450/commits/6f5fc788f264e56698ebc825817079ff94eae1f7. Thoughts?

Reply:
As I wrote internally - I am not sure we need to require users pass in `Bookkeeper`. All `Bookkeper` needs to do is let `RealMarket` know it should eagerly resolve conflicts. Initially I thought we would want to defer implementation to users. But with `Bookkeeping` mental model, I suspect most users will not care to customize. We can bind `Bookkeeper` to a disk cache or DB ourselves under the hood and make it default implementation. Thoughts on making `Bookkeeper` optional?

Reply:
Let's revisit this after getting this PR merged (#452)